accepted/2020/binary-data/binary-data.md

```diff
@@ -0,0 +1,195 @@
+data.SetBytes(...);
+```
+
+This would also allow us to use regular extension methods to split out JSON
+serialization.
+
+We need to balance layering concerns with usability. The primary goal of this
+type is to provide usability via conveniences. If we don't make the type super
+usable, the feature has failed. While the design with extension methods feels
+quite appealing we need to verify that it realizes these goal. If we can't, we
+should go back to the design of using constructors and factory methods, even if
+that reduces our flexibility.
+
+### API design with extension methods
+
+```C#
+namespace System
+{
+    public readonly struct BinaryData
+    {
+        public void SetBytes(ReadOnlyMemory<byte> data);
+        public void SetBytes(ReadOnlySpan<byte> data);
+        public void SetBytes(byte[] data);
+        public void SetStream(Stream stream);
+        public static Task SetStreamAsync(Stream stream, CancellationToken cancellationToken = default);
+        public void SetString(string data);
+        public static implicit operator ReadOnlySpan<byte>(BinaryData data);
+        public static implicit operator ReadOnlyMemory<byte>(BinaryData data);
+        public ReadOnlyMemory<byte> ToBytes();
+        public Stream ToStream();
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public override bool Equals(object? obj);
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public override int GetHashCode();
+        public override string ToString();
+    }
+}
+namespace System.Text.Json
+{
+    public static class BinaryDataJsonExtensions
+    {
+        public static Task SetObjectAsJsonAsync<T>(this BinaryData data, T jsonSerializable, CancellationToken cancellationToken = default);
+        public static void SetObjectAsJson<T>(this BinaryData data, T jsonSerializable, CancellationToken cancellationToken = default);
+        public static T ToObjectFromJson<T>(this BinaryData data, CancellationToken cancellationToken = default);
+        public static ValueTask<T> ToObjectFromJsonAsync<T>(this BinaryData data, CancellationToken cancellationToken = default);
+    }
+}
+```
+
+### API with constructors and factory methods
+
+```C#
+namespace System
+{
+    public readonly struct BinaryData
+    {
+        public BinaryData(ReadOnlySpan<byte> data);
+        public BinaryData(byte[] data);
+        public BinaryData(object jsonSerializable, Type? type = null);
+        public BinaryData(ReadOnlyMemory<byte> data);
+        public BinaryData(string data);
+        public static BinaryData FromBytes(ReadOnlyMemory<byte> data);
+        public static BinaryData FromBytes(ReadOnlySpan<byte> data);
+        public static BinaryData FromBytes(byte[] data);
+        public static BinaryData FromObject<T>(T jsonSerializable, CancellationToken cancellationToken = default);
+        public static Task<BinaryData> FromObjectAsync<T>(T jsonSerializable, CancellationToken cancellationToken = default);
```

Comment:
In the ctors/factory version we probably still want "Json" in the name of the serialize/deserialize methods.  If nothing else, it sets up the pattern for the next data format type.